core/service/test/src/lib.rs

```diff
@@ -51,15 +54,26 @@ struct TestNet<F: ServiceFactory> {
 
 impl<F: ServiceFactory> TestNet<F> {
 	pub fn run_until_all_full<P: Send + Sync + Fn(u32, &F::FullService) -> bool + 'static>(&mut self, predicate: P) {
+		let begin = Instant::now();
 		let full_nodes = self.full_nodes.clone();
 		let interval = Interval::new_interval(Duration::from_millis(100)).map_err(|_| ()).for_each(move |_| {
+			if Instant::now() - begin > MAX_WAIT_TIME {
```

Comment:
Could we make this configurable and please use https://docs.rs/tokio/0.1.21/tokio/prelude/trait.StreamExt.html#method.timeout

Reply:
Sure. What do you mean by configurable?

Reply:
`timeout` won't work because it is per-item && there are no actually items in our stream. The item is just check if test has completed

Reply:
I send the incorrect one, `for_each` is a future, so this should be used: https://docs.rs/tokio/0.1.21/tokio/prelude/trait.FutureExt.html#method.timeout

Reply:
Sorry, still missing the idea :) The `run_until_all_full` is simply `Interval` (stream) which for every item runs almost-instant future - which just checks if predicate returns `true` for all nodes. The actual work is competed outside of this check. The method you're proposing will early-return from future with timeout error if future runs for too long. But that's not our case - as I said before, the future is almost instant => timeout will never be fired. I guess we could use `Stream::take_while`, but it is quite similar to current implementation.

Reply:
`for_each` converts a `Stream` into a `Future`: https://docs.rs/futures/0.1.27/futures/stream/struct.ForEach.html
So, making this future timeout, will still execute the inner stream and the checks are done as usual, but if the future hits the limit, it is aborted. This makes the code cleaner and does not require you do replicate your check.